pretyping/pretyping.ml

```diff
@@ -478,39 +478,138 @@ let mark_obligation_evar sigma k evc =
-  let inh_conv_coerce_to_tycon ?loc = inh_conv_coerce_to_tycon ?loc ~program_mode resolve_tc in
-  let pretype_type = pretype_type ~program_mode ~poly resolve_tc in
-  let pretype = pretype ~program_mode ~poly resolve_tc in
-  let open Context.Rel.Declaration in
+(** Tie the loop *)
+let eval_pretyper self ~program_mode ~poly resolve_tc tycon env sigma t =
   let loc = t.CAst.loc in
   match DAst.get t with
-  | GRef (ref,u) ->
+  | GRef (ref,u) -> self.pretype_ref self (ref, u) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GVar id -> self.pretype_var self id ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GEvar (evk, args) -> self.pretype_evar self (evk, args) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GPatVar knd -> self.pretype_patvar self knd ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GApp (c, args) -> self.pretype_app self (c, args) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GLambda (na, bk, t, c) -> self.pretype_lambda self (na, bk, t, c) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GProd (na, bk, t, c) -> self.pretype_prod self (na, bk, t, c) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GLetIn (na, b, t, c) -> self.pretype_letin self (na, b, t, c) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GCases (st, c, tm, cl) -> self.pretype_cases self (st, c, tm, cl) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GLetTuple (na, b, t, c) -> self.pretype_lettuple self (na, b, t, c) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GIf (c, r, t1, t2) -> self.pretype_if self (c, r, t1, t2) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GRec (knd, nas, decl, c, t) -> self.pretype_rec self (knd, nas, decl, c, t) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GSort s -> self.pretype_sort self s ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GHole (knd, nam, arg) -> self.pretype_hole self (knd, nam, arg) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GCast (c, t) -> self.pretype_cast self (c, t) ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GInt n -> self.pretype_int self n ?loc ~program_mode ~poly resolve_tc tycon env sigma
+  | GFloat f -> self.pretype_float self f ?loc ~program_mode ~poly resolve_tc tycon env sigma
+
+let eval_type_pretyper self ~program_mode ~poly resolve_tc tycon env sigma t =
+  self.pretype_type self t ~program_mode ~poly resolve_tc tycon env sigma
+
+let pretype_instance self ~program_mode ~poly resolve_tc env sigma loc hyps evk update =
+  let f decl (subst,update,sigma) =
+    let id = NamedDecl.get_id decl in
+    let b = Option.map (replace_vars subst) (NamedDecl.get_value decl) in
+    let t = replace_vars subst (NamedDecl.get_type decl) in
+    let check_body sigma id c =
+      match b, c with
+      | Some b, Some c ->
+         if not (is_conv !!env sigma b c) then
```

Comment:
Nit: maybe a good opportunity to improve indentation here?

Reply:
I am leaving the code that way to minimize the diff, but we could do that in another patch though.

Reply:
The thing is that already the diff is there as this has moved.

Reply:
What do you have in mind as a result?

Reply:
It seems the code uses too much space after the cases, as for example in
```
      | Some b, None ->
￼           user_err ?loc  (str "Cannot interpret " ++
￼             pr_existential_key sigma evk ++
```
but anyways I'm just nitpicking; I always apply `ocp-indent` from emacs and it takes care of that stuff.